src/Compilers/CSharp/Test/Emit/Emit/AsyncMainEntryPointTests.cs

```diff
@@ -0,0 +1,386 @@
+﻿using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Emit
+{
+    public class AsyncMainEntryPointTests : CompilingTestBase
+    {
+        [Fact]
+        public void MainCanBeAsyncWithArgs()
+        {
+            var origSource = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+    }
+}";
+            var sources = new string[] { origSource, origSource.Replace("async ", "").Replace("await", "return") };
+            foreach (var source in sources)
+            {
+                var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+                compilation.VerifyDiagnostics();
+                var entry = compilation.GetEntryPoint(CancellationToken.None);
+                Assert.NotNull(entry);
+                Assert.Equal("System.Threading.Tasks.Task A.Main(System.String[] args)", entry.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void MainCantBeAsyncWithRefTask()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    static ref Task Main(string[] args)
+    {
+        throw new System.Exception();
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+            compilation.VerifyDiagnostics(
+                // (6,21): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point
+                //     static ref Task Main(string[] args)
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("A.Main(string[])").WithLocation(6, 21),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+        }
+
+        [Fact]
+        public void MainCantBeAsyncWithArgs_CSharp7()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7));
+            compilation.VerifyDiagnostics(
+                // (6,5): error CS8107: Feature 'async main' is not available in C# 7. Please use language version 7.1 or greater.
+                //     async static Task Main(string[] args)
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task Main(string[] args)
```

Comment:
>Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task Main(string[] args) [](start = 16, length = 95)

I believe we should change location of this error to be just the return type rather than the entire method with the body.